### PR TITLE
R-Py parity for ml-regression page

### PR DESCRIPTION
**Pull request by kvdesai**

### CODE REVIEW

File: r/2021-07-08-ml-regression.Rmd [hunk 1]
```diff
@@ -0,0 +1,312 @@
+<!-- #region -->
+This page shows how to use Plotly charts for displaying various types of regression models, starting from simple models like Linear Regression and progressively move towards models like Decision Tree and Polynomial Features. We highlight various capabilities of plotly, such as comparative analysis of the same model with different parameters, displaying Latex, and [surface plots](https://plotly.com/r/3d-surface-plots/) for 3D data.
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:2]:
```suggestion
This page shows how to use Plotly charts for displaying various types of regression models, starting from simple models like [Linear Regression](https://parsnip.tidymodels.org/reference/linear_reg.html) and progressively move towards models like Decision Tree and Polynomial Features. We highlight various capabilities of plotly, such as comparative analysis of the same model with different parameters, displaying Latex, and [surface plots](https://plotly.com/r/3d-surface-plots/) for 3D data.
```

Can we also add the in-line hyperlinks to "Decision Tree" and "Polynomial Features" for parity? In the Python docs we have hyperlinks for the `sk-learn` docs, maybe we can have something similar with the `parsnip` or `tidymodels` references.

File: r/2021-07-08-ml-regression.Rmd [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+
+We will use [tidymodels](https://tidymodels.tidymodels.org/) to split and preprocess our data and train various regression models. Tidymodels is a popular Machine Learning (ML) library in R that is compatible with the "tidyverse" concepts, and offers various tools for creating and training ML algorithms, feature engineering, data cleaning, and evaluating and testing models. It is the next-gen version of the popular [caret](http://topepo.github.io/caret/index.html) library in R.
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:4]:
```suggestion
We will use [tidymodels](https://tidymodels.tidymodels.org/) to split and preprocess our data and train various regression models. Tidymodels is a popular Machine Learning (ML) library in R that is compatible with the "tidyverse" concepts, and offers various tools for creating and training ML algorithms, feature engineering, data cleaning, and evaluating and testing models. It is the next-gen version of the popular [caret](http://topepo.github.io/caret/index.html) library for R.
```

File: r/2021-07-08-ml-regression.Rmd [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+
+<!-- #endregion -->
+
+## Basic linear regression plots
+
+In this section, we show you how to apply a simple regression model for predicting tips a server will receive based on various client attributes (such as sex, time of the week, and whether they are a smoker).
+
+We will be using the [Linear Regression][lr], which is a simple model that fit an intercept (the mean tip received by a server), and add a slope for each feature we use, such as the value of the total bill.
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:12]:
```suggestion
We will be using the [Linear Regression][lr], which is a simple model that fits an intercept (the mean tip received by a server), and adds a slope for each feature we use, such as the value of the total bill.
```

Note - Instead of `[lr]` I think we want to be consistent with our hyperlinks, I would just make them all Markdown-style.

File: r/2021-07-08-ml-regression.Rmd [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+
+x_range <- seq(min(X), max(X), length.out = 100)
+x_range <- matrix(x_range, nrow=100, ncol=1)
+x_range <- data.frame(x_range)
+colnames(x_range) <- c('total_bill')
+
+y_range <- lm_model %>% predict(x_range) 
+
+colnames(y_range) <- c('tip')
+xy <- data.frame(x_range, y_range)
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:41]: One suggestion here is to rename the `dataframe` variable to `df` for this example and other ones (unless there's multiple dataframes).

File: r/2021-07-08-ml-regression.Rmd [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+colnames(y_range) <- c('tip')
+xy <- data.frame(x_range, y_range) 
+
+fig <- plot_ly(tips, x = ~total_bill, y = ~tip, type = 'scatter', alpha = 0.65, mode = 'markers', name = 'Tips')
+fig <- fig %>% add_trace(data = xy, x = ~total_bill, y = ~tip, name = 'Regression Fit', mode = 'lines', alpha = 1)
+fig
+```
+## Model generalization on unseen data
+
+With `add_trace()`, you can easily color your plot based on a predefined data split. By coloring the training and the testing data points with different colors, you can easily see if whether the model generalizes well to the test data or not.
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:49]:
```suggestion
With `add_trace()`, you can easily color your plot based on a predefined data split. By coloring the training and the testing data points with different colors, you can easily see if the model generalizes well to the test data or not.
```

File: r/2021-07-08-ml-regression.Rmd [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+fig
+```
+
+## Comparing different kNN models parameters
+
+In addition to linear regression, it's possible to fit the same data using [k-Nearest Neighbors][knn]. When you perform a prediction on a new sample, this model either takes the weighted or un-weighted average of the neighbors. In order to see the difference between those two averaging options, we train a kNN model with both of those parameters, and we plot them in the same way as the previous graph.
+
+Notice how we can combine scatter points with lines using Plotly. You can learn more about [multiple chart types](https://plotly.com/r/graphing-multiple-chart-types/).
+
+[knn]: https://scikit-learn.org/stable/modules/generated/sklearn.neighbors.KNeighborsRegressor.html
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:96]: We'll want to replace this with the R example instead.

File: r/2021-07-08-ml-regression.Rmd [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+X <- tips$total_bill
+
+knn_dist <- nearest_neighbor(neighbors = 10, weight_func = 'inv') %>% 
+  set_engine('kknn') %>% 
+  set_mode('regression') %>%
+  fit(tip ~ total_bill, data = tips)
+knn_uni <- nearest_neighbor(neighbors = 10, weight_func = 'rectangular') %>% 
+  set_engine('kknn') %>% 
+  set_mode('regression') %>%
+  fit(tip ~ total_bill, data = tips)
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:116]:
```suggestion
# Model #1
knn_dist <- nearest_neighbor(neighbors = 10, weight_func = 'inv') %>% 
  set_engine('kknn') %>% 
  set_mode('regression') %>%
  fit(tip ~ total_bill, data = tips)
  
# Model #2
knn_uni <- nearest_neighbor(neighbors = 10, weight_func = 'rectangular') %>% 
  set_engine('kknn') %>% 
  set_mode('regression') %>%
  fit(tip ~ total_bill, data = tips) 
```

File: r/2021-07-08-ml-regression.Rmd [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+fig <- plot_ly(tips, type = 'scatter', mode = 'markers', colors = c("#FF7F50", "#6495ED")) %>% 
+  add_trace(data = tips, x = ~total_bill, y = ~tip, type = 'scatter', mode = 'markers', color = ~sex, alpha = 0.65) %>% 
+  add_trace(data = xy, x = ~total_bill, y = ~dist, name = 'Weights: Distance', mode = 'lines', alpha = 1) %>%
+  add_trace(data = xy, x = ~total_bill, y = ~uni, name = 'Weights: Uniform', mode = 'lines', alpha = 1) 
+fig
+```
+
+## 3D regression surface with `mesh3d` and `add_surface`
+
+Visualize the decision plane of your model whenever you have more than one variable in your input data. Here, we will use [`svm_rbf`](https://parsnip.tidymodels.org/reference/svm_rbf.html) with [`kernlab`](https://cran.r-project.org/web/packages/kernlab/index.html) engine in `regression` mode. For generating the 2D mesh on the surface, we use the package [`pracma`](https://cran.r-project.org/web/packages/pracma/index.html)
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:141]:
```suggestion
Visualize the decision plane of your model whenever you have more than one variable in your input data. Here, we will use [`svm_rbf`](https://parsnip.tidymodels.org/reference/svm_rbf.html) with [`kernlab`](https://cran.r-project.org/web/packages/kernlab/index.html) engine in `regression` mode. For generating the 2D mesh on the surface, we use the  [`pracma`](https://cran.r-project.org/web/packages/pracma/index.html) package.
```

File: r/2021-07-08-ml-regression.Rmd [hunk 9]
```diff
@@ -0,0 +1,312 @@
...
+xx1 <- matrix(xx, length(xx), 1)
+yy1 <- matrix(yy, length(yy), 1)
+final <- cbind(xx1, yy1)
+pred <- model %>%
+  predict(final)
+
+pred <- pred$.pred
+pred <- matrix(pred, dim_val[1], dim_val[2])
+
+dim(pred)
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:181]: Do we need this? If not, I would :scissors:

File: r/2021-07-08-ml-regression.Rmd [hunk 10]
```diff
@@ -0,0 +1,312 @@
...
+
+hist_top <- ggplot(iris,aes(x=Petal.Width)) + 
+  geom_histogram(data=subset(iris,split == 'train'),fill = "red", alpha = 0.2, bins = 6) +
+  geom_histogram(data=subset(iris,split == 'test'),fill = "blue", alpha = 0.2, bins = 6) +
+  theme(axis.title.y=element_blank(),axis.text.y=element_blank(),axis.ticks.y=element_blank())
+hist_top <- ggplotly(p = hist_top)
+
+scatter <- ggplot(iris, aes(x = Petal.Width, y = prediction, color = split)) +
+  geom_point() + 
+  geom_smooth(method=lm, se=FALSE)
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:235]:
```suggestion
  geom_smooth(formula=y ~ x, method=lm, se=FALSE) 
```

This will avoid the unnecessary warning.

File: r/2021-07-08-ml-regression.Rmd [hunk 11]
```diff
@@ -0,0 +1,312 @@
...
+  predict(X) 
+colnames(prediction) <- c('prediction')
+iris = cbind(iris, prediction)
+residual <- prediction - iris$Petal.Width
+colnames(residual) <- c('residual')
+iris = cbind(iris, residual)
+
+scatter <- ggplot(iris, aes(x = prediction, y = residual, color = split)) + 
+  geom_point() + 
+  geom_smooth(method=lm, se=FALSE)
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:297]:
```suggestion
  geom_smooth(formula=y ~ x, method=lm, se=FALSE) 
```

File: r/2021-07-08-ml-regression.Rmd [hunk 12]
```diff
@@ -0,0 +1,312 @@
...
+  scatter,
+  hist_right,
+  nrows = 2, heights = c(0.2, 0.8), widths = c(0.8, 0.2), margin = 0,
+  shareX = TRUE, shareY = TRUE, titleX = TRUE, titleY = TRUE
+)
+layout(s, showlegend = FALSE)
+
+```
+## Residual plots
+Just like prediction error plots, it's easy to visualize your prediction residuals in just a few lines of codes using `ggplotly` built-in capabilities.
```
Comment (HammadTheOne) [inline r/2021-07-08-ml-regression.Rmd:257]:
```suggestion
Just like prediction error plots, it's easy to visualize your prediction residuals in just a few lines of codes using `ggplotly` and `tidymodels` capabilities.
```

**Review by HammadTheOne — COMMENTED** [review]
Hey @kvdesai , these examples look great! They're fast and fluid, and I like the syntax used. I was able to run all of them by copying and pasting the code into my local environment and running each example script. The `knitted` html also looks good. 

I've made a few suggestions/comments that would bring it across the finish line. Along with that, I would suggest maybe just taking one more pass over the `knitr` generated HTML and addressing any of the non-blocking warnings that appear (eg. the `geom_smooth` formula ones). This will prevent any panic from inexperienced users who might not know what's causing the warning or what they mean. 


Other than that, it looks fantastic. I can approve it once those minor changes have been addressed.

**Comment by kvdesai** [comment]
@HammadTheOne - thanks for the review and insightful comments. I have addressed them. I also added one more snippet of the plot. Please verify and let me know if any other change is needed.

**Review by HammadTheOne — APPROVED** [review]
Looks good! Thanks for making the requested changes. 

 I notice that one of the CI tests is failing now though (possibly related to 2016-06-17-3d-tri-surf.Rmd), so once that's fixed we can go ahead and merge it.

:woman_dancing:
